### ISSUE TITLE
Improve last summary flow

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#9946]
## Goal
Reduce number of trailing ops in files at rest.
Ideal (unachievable outcome at the moment) - there are no trailing ops in files at rest.

## Suggested work
1. Pass SummarizerStopReason to shouldRunLastSummary()
2. Add telemetry event in shouldRunLastSummary() to log opsSinceLastAck & this.minOpsForAttemptOnClose & SummarizerStopReason
3. Make shouldRunLastSummary() return false if reason is not "parentNotConnected".
4. Change default value for minOpsForAttemptOnClose from 50 to 10
5. Change logic in getShouldSummarizeState() - reorder first two cases (parentNotConnected & parentShouldNotSummarize)
6. RunningSummarizer.trySummarize() should only do one iteration (see attempts array) when summarizeReason == "lastSummary"
7. Expose minOpsForAttemptOnClose on ISummaryConfiguration (and move default to DefaultSummaryConfiguration. Such settings should not be buried deep in the code, and specific scenario (i.e., app) should be able to change them easily
   - I'd prefer it to be a separate PR, this is lower priority compared to other items.

## Description

There are only 2.5K last summaries out of 6M of summaries - that's very low number!

> union Office_Fluid_FluidRuntime_*
| where Data_eventName == "fluid:telemetry:Summarizer:Running:SummaryOp"
| summarize count() by Data_summarizeReason

Last summary is the process that runs to summarize a file when main client is closed. It was originally built as a recovery mechanism for corrupted files (files that have over 10K unsummarized ops, and thus can't be edited unless new summary is posted).

At the same time, there are 3.6M summary sessions, some of them do not log the following event (as tab is closed), but most do:
> union Office_Fluid_FluidRuntime_*
| where Data_eventName contains "StoppingSummarizer"
| where Data_clientType != "interactive"
| summarize count() by Data_reason

Data_reason | count_
-- | --
parentShouldNotSummarize | 2,119,619
parentNotConnected | 299,577
summarizerClientDisconnected | 45,086
failToSummarize | 58
summarizerException | 5

Top 2 lines can run last summary (allowLastSummary = true in waiStop()) today but they do not due to too high minOpsForAttemptOnClose.
That's a bummer as we can leverage that to reduce number of cases where we have trailing ops!

reason = "summarizerClientDisconnected" should translate to allowLastSummary = false, the bottom 2 reasons probably should never try to do last summary, that said not sure how much we care. parentShouldNotSummarize is questionable whether we want to run a summary, as it means there is another summarizer client that is booting in parallel, so running last summary here risks having two summarizers, and thus we should avoid running last summary.

But parentNotConnected indicates that parent is disconnected. In original design that was equivalent to parentShouldNotSummarize, but with recent Paul's changes a summarizer client that stays afloat is considered to be elected, and should proceed to last summary.
For that reason, I believe the order of getShouldSummarizeState() should be changed - we should first check if client is condition for parentShouldNotSummarize is satisfied - this is stronger signal. We should also likely rename it, as it has less to do with parent nowadays. This change (reorder) will make this logic independent on how decisions for elected summarizer is made (as with today's logic we are essentially overwriting such logic in getShouldSummarizeState() by saying that parent connectivity is included in this logic, while it might be not true).

We might go with allow-list here, and shouldRunLastSummary() can only return true for first two reasons.

**Comment by agarwal-navin** [comment]
From my comment here - https://github.com/microsoft/FluidFramework/issues/8995#issuecomment-1101799449.
I think the last summary by a summarizer should only do 1 (or max 2) attempts before it shuts down. The problem is that when the summarize is in a state where its summaries are failing (like the issue in #8995), it will keep retrying for a long time with delays in between and fail every time. As per the analysis in that issue, a new summarizer will be created but the current one does not go away because it keeps on retrying to summarize.

**Comment by vladsud** [comment]
Agree. I've added item # 6 in the description above to cover it - please take a look.

**Comment by agarwal-navin** [comment]
For 4), do we need a limit on number of ops since last summary? Shouldn't we summarize even if there is one non-system op - kind of like saving before existing?

**Comment by vladsud** [comment]
@kian-thompson, do we have an issue to change 50 -> 10? (# 4 in list above)?
Same for # 7 (not sure if it was part of Nicholas change - need to expose it properly in config)

**Comment by kian-thompson** [comment]
> @kian-thompson, do we have an issue to change 50 -> 10? (# 4 in list above)? Same for # 7 (not sure if it was part of Nicholas change - need to expose it properly in config)

Yes, I accidentally closed this item. My apologies.

**Comment by NicholasCouri** [comment]
I exposed #7 -we just need to update it…
